pkg/scheduler/plugins/reservation/transformer.go

```diff
@@ -22,210 +22,361 @@ import (
-	return preparePreFilterPod(pod), true
-}
+	if !reservationutil.IsReservePod(pod) {
+		for nodeName, reservationInfos := range state.matched {
+			nodeInfo, err := pl.handle.SnapshotSharedLister().NodeInfos().Get(nodeName)
+			if err != nil {
+				continue
+			}
+			if nodeInfo.Node() == nil {
+				continue
+			}
+
+			// NOTE: After PreFilter, all reserved resources need to be returned to the corresponding NodeInfo,
+			// and each plugin is triggered to adjust its own StateData through RemovePod, RemoveReservation and AddPodInReservation,
+			// and adjust the state data related to Reservation and Pod
+			if err := pl.restoreMatchedReservations(cycleState, pod, nodeInfo, reservationInfos, true); err != nil {
+				return err
+			}
+		}
+	}
 
-func (p *Plugin) AfterPreFilter(handle frameworkext.ExtendedHandle, cycleState *framework.CycleState, pod *corev1.Pod) error {
 	return nil
 }
 
-func (p *Plugin) BeforeFilter(handle frameworkext.ExtendedHandle, cycleState *framework.CycleState, pod *corev1.Pod, nodeInfo *framework.NodeInfo) (*corev1.Pod, *framework.NodeInfo, bool) {
-	// do not transformer if not reserve state (where we should check if pod match any reservation on node)
-	if reservationutil.IsReservePod(pod) {
-		return nil, nil, false
-	}
-	// abort if node is not found
+func (pl *Plugin) BeforeFilter(handle frameworkext.ExtendedHandle, cycleState *framework.CycleState, pod *corev1.Pod, nodeInfo *framework.NodeInfo) (*corev1.Pod, *framework.NodeInfo, bool) {
 	node := nodeInfo.Node()
 	if node == nil {
 		return nil, nil, false
 	}
+	// Generation equals -1 means that the Reservation return logic has been executed once, but it should not happen.
+	if nodeInfo.Generation == -1 {
+		return nil, nil, false
+	}
 
-	// only continue if matchedCache is prepared in BeforePreFilter (i.e. find any reservation matched)
-	state := getPreFilterState(cycleState)
-	if state == nil {
+	state := getStateData(cycleState)
+	matchedReservationInfos := state.matched[node.Name]
+	unmatchedReservationInfos := state.unmatched[node.Name]
+	if len(matchedReservationInfos) == 0 && len(unmatchedReservationInfos) == 0 {
 		return nil, nil, false
 	}
 
-	allocatedResource, ok := state.allocatedResources[node.Name]
-	// skip when no reservation matched on this node
-	rOnNode := state.matchedCache.GetOnNode(node.Name)
-	if len(rOnNode) <= 0 && !ok {
+	nodeInfo = nodeInfo.Clone()
+	if err := pl.restoreUnmatchedReservations(cycleState, pod, nodeInfo, unmatchedReservationInfos, false); err != nil {
 		return nil, nil, false
 	}
 
-	klog.V(5).InfoS("BeforeFilter get reservation matched on node",
-		"pod", klog.KObj(pod), "node", node.Name, "count", len(rOnNode))
-	// fix-up reserved resources and ports
-	return pod, prepareFilterNodeInfo(pod, nodeInfo, rOnNode, allocatedResource), true
+	if !reservationutil.IsReservePod(pod) {
+		if err := pl.restoreMatchedReservations(cycleState, pod, nodeInfo, matchedReservationInfos, false); err != nil {
```

Comment:
1. Why does resource returning logic executed twice in beforeFilter and afterPreFilter?
2. Why afterPreFilter not beforePreFilter

Reply:
1. In Filter phase, the parameter `NodeInfo` comes from the internal cache of `GenericScheduler` that is original data, this data also retains Reserve Pod data, so it needs to be processed again before executing Filter.
2. Processing in AfterPreFilter, relatively speaking, the implementation will be unified, and both upstream plug-ins and internal plug-ins can be considered, and the implementation is simpler and easier to maintain